### PR TITLE
fix: improve context compaction with preserved recent context

### PR DESCRIPTION
**Pull request by nickbaumann98**
### Related Issue

Addresses user feedback about auto-compact being disruptive and losing conversational continuity.

### Description

This PR improves the auto-compact feature to be less disruptive while maintaining effective context management. The current system compresses the entire conversation history, causing users to lose track of recent context and creating jarring breaks in task continuity.

**Key improvements:**
- **Smarter triggers**: Now uses "75% OR 20K headroom" rule (whichever comes first), adapting intelligently to different model context sizes
- **Preserves recent context**: Keeps the last 8 user-assistant exchanges intact, maintaining conversation flow
- **Better summarization**: Shifted from comprehensive documentation to narrative-focused summaries that capture key moments
- **Edge case handling**: Failsafe check skips compaction if middle section is too small, letting legacy truncation handle edge cases instead
- **Prevents premature re-compaction**: Added warning to continuation prompt reminding AI not to compact unless at 75%+

### Test Procedure

- Manually tested compaction triggering at various context levels
- Verified recent messages are preserved after compaction
- Confirmed failsafe prevents compaction on short conversations
- Ran `npm test` - all tests passing
- Ran `npm run format:fix` - code formatted

### Type of Change

- [x] 🐛 Bug fix (non-breaking change which fixes an issue)
- [x] ✨ New feature (non-breaking change which adds functionality)

### Pre-flight Checklist

- [x] Changes are limited to a single feature, bugfix or chore (split larger changes into separate PRs)
- [x] Tests are passing (`npm test`) and code is formatted and linted (`npm run format && npm run lint`)
- [x] I have created a changeset using `npm run changeset` (required for user-facing changes)
- [x] I have reviewed [contributor guidelines](https://github.com/cline/cline/blob/main/CONTRIBUTING.md)

### Additional Notes

This maintains backward compatibility with the existing legacy truncation system, which serves as a fallback for edge cases. The improved system only applies to next-gen models (Claude 4, Gemini 2.5, GPT-5, Grok 4) that support auto-condense.

----

> [!IMPORTANT]
> Improves context compaction by preserving recent messages and implementing smarter triggers, with changes in context window management and summarization techniques.
> 
>   - **Behavior**:
>     - Implements "75% OR 20K headroom" rule in `getContextWindowInfo()` in `context-window-utils.ts` for context window management.
>     - Preserves last 8 user-assistant exchanges in `Task` class in `index.ts` to maintain conversation flow.
>     - Adds failsafe in `Task` class to skip compaction if middle section is too small.
>   - **Summarization**:
>     - Updates `summarizeTask()` in `contextManagement.ts` to focus on narrative summaries, preserving recent messages.
>     - Adds guidelines to ensure summaries capture key moments and technical details.
>   - **Misc**:
>     - Adds warning in `continuationPrompt()` in `contextManagement.ts` to prevent premature re-compaction.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=cline%2Fcline&utm_source=github&utm_medium=referral)<sup> for ec24ba962da544df6c4cbdb0dc9b98da0491da2a. You can [customize](https://app.ellipsis.dev/cline/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: src/core/prompts/contextManagement.ts [hunk 1]
```diff
@@ -8,40 +8,48 @@ export const summarizeTask = (focusChainSettings?: { enabled: boolean }, cwd?: s
...
-8. Next Step: List the next step that you will take that is related to the most recent work you were doing. IMPORTANT: ensure that this step is DIRECTLY in line with the user's explicit requests, and the task you were working on immediately before this summary request. If your last task was concluded, then only list next steps if they are explicitly in line with the users request. Do not start on tangential requests without confirming with the user first.
-                     If there is a next step, include direct quotes from the most recent conversation showing exactly what task you were working on and where you left off. This should be verbatim to ensure there's no drift in task interpretation.
-9. Required Files: List the most important files needed for continuing the work you laid out in Next Step. This is optional and if no files are required or there is no next step then simply don't include this section. List each file path on a new line starting with "- " such as: - src/main.js. List the files from most important to least important. You must list the minimum number of files necessary to continue with the task.
```
Comment (0xToshii) [inline src/core/prompts/contextManagement.ts:42]: cursory read but i think this keeps both the message for creating the summary along with the summary itself which is not ideal. also the conversation history is linear so this effectively puts a summary of previous message after the future messages that occur after that summary which might confuse the model

possibly with keeping these messages we might be able to remove the required files section, but we need to test this to see whether this is the case. same goes for the attempt completion

**Comment by saoudrizwan** [comment]
Closing for now unless we wanna re-hash and re-open @nickbaumann98
